pkg/controller/warehouses/warehouses.go

```diff
@@ -646,6 +690,61 @@ func validateDiscoveredArtifacts(
 	return true
 }
 
+// freightCreationCriteriaSatisfied evaluates the freight creation criteria
+// expression, if defined, and returns true if the expression is satisfied,
+// no expression is defined, or no discovered artifacts are present.
+// A non-nil error is returned if there was an error evaluating the expression.
+func freightCreationCriteriaSatisfied(
+	ctx context.Context,
+	wh *kargoapi.Warehouse,
+	artifacts *kargoapi.DiscoveredArtifacts,
+) (bool, error) {
+	logger := logging.LoggerFromContext(ctx)
+
+	if wh.Spec.FreightCreationCriteria == nil {
+		logger.Debug("no freight creation criteria")
+		return true, nil
+	}
+
+	expression := strings.TrimSpace(wh.Spec.FreightCreationCriteria.Expression)
+	if expression == "" {
+		logger.Debug("no freight creation criteria expression")
+		return true, nil
+	}
+
+	if artifacts == nil || (len(artifacts.Git) == 0 && len(artifacts.Images) == 0 && len(artifacts.Charts) == 0) {
+		logger.Debug("no artifacts discovered")
+		return true, nil
+	}
+
+	logger.WithValues("criteriaExpression", expression)
```

Comment:
This line is having no effect.

It would need to be:

```
logger = logger.WithValues("criteriaExpression", expression)
Comment
```

That said, there's only one log message that _possibly_ written after this point in the code, so you could just add this key/val pair to the log output on line +732.

Somewhat related, and not isolated to this PR, I think we're possibly adding too much noise to the logs lately. It's not a hill I'd die on, but you might consider knocking these messages down to trace level.